### PR TITLE
DXSM ambiguous sequences

### PR DESCRIPTION
**Pull request by willdumm**
This PR addresses https://github.com/matsengrp/dnsm-experiments-1/issues/36 and may eventually come with a companion PR in that repository.

### CODE REVIEW

**Comment by matsen** [comment]
This is GREAT! Thank you for trying this out and coming up with these corner cases.

The overall strategy is that we should take the simplest way out. 

If there is an N anywhere in the codon that whole codon should be masked out, even if it's synonymous variation. I believe that the data sets that @ksung25 has put together are constrained so that if there is an N in some sequence in a site then there is an N for all sequences at that site. (This answers "Does the construction of masks need to consider ambiguities in the child sequence?") We could do further preprocessing so that the whole codon is masked out in that case, or we could do it on our side.

We should throw out any PCP such that the only difference between the parent and child are mutations in N-containing codons. Again, this could be a pre-processing step or on our side. 

> When datasets are split, we use the constructor from scratch. This can behave oddly because we re-compute the mask according to the max length of the subset of sequences, but nt_csps length is determined by the max length of a sequence in the whole original dataset. we could have alternative constructor that allows us to fix the mask to a subset of the existing
masks...

I see the problem, don't quite get your proposed solution, but trust you to take this forward. What if we just supply the length in the constructor (optionally)?

> There are some places where we re-compute masks and aa sequences from the nucleotide string. It seems that we should refactor these to make use of the already-computed masks and aa string that we have stored in the Dataset. An example is build_selection_matrix_from_parent

Seems reasonable. Not sure I know why I didn't do that? If you wanted to test this out before the rest of the changes you could do that in a separate branch that we could merge before doing further work on this branch. It'd be nice to test that change in isolation.

**Comment by ksung25** [comment]
> I believe that the data sets that @ksung25 has put together are constrained so that if there is an N in some sequence in a site then there is an N for all sequences at that site. (This answers "Does the construction of masks need to consider ambiguities in the child sequence?")
> 
Just want to elaborate that PCP files with `MASKED` in the name will have the constraint Erick mentioned. One note is that the Wyatt/10x dataset do not have N at all, so those files won't have `MASKED` labels because nothing changes.

File: netam/dasm.py [hunk 1]
```diff
@@ -36,8 +36,10 @@ def update_neutral_probs(self):
 
             mut_probs = 1.0 - torch.exp(-branch_length * nt_rates[:parent_len])
             nt_csps = nt_csps[:parent_len, :]
-            molevol.check_csps(parent_idxs, nt_csps)
+            nt_mask = mask.repeat_interleave(3)[: len(nt_parent)]
+            molevol.check_csps(parent_idxs[nt_mask], nt_csps[: len(nt_parent)][nt_mask])
 
+            # TODO don't we need to pass multihit model in here?
```
Comment (willdumm) [inline netam/dasm.py:42]: We do pass in the multihit model in the analogous spot in `dnsm.py`...

File: netam/dasm.py [hunk 2]
```diff
@@ -196,11 +198,17 @@ def loss_of_batch(self, batch):
...
+
+        Values at ambiguous sites are meaningless.
+        """
         # This is simpler than the equivalent in dnsm.py because we get the selection
         # matrix directly. Note that selection_factors_of_aa_str does the exponentiation
         # so this indeed gives us the selection factors, not the log selection factors.
         parent = sequences.translate_sequence(parent)
         per_aa_selection_factors = self.model.selection_factors_of_aa_str(parent)
+
+        parent = parent.replace("X", "A")
```
Comment (willdumm) [inline netam/dasm.py:211]: Noted in docstring that these sites are nonsense. They are masked where this is used in `_find_optimal_branch_length`.

File: netam/dnsm.py [hunk 3]
```diff
@@ -45,9 +45,13 @@ def update_neutral_probs(self):
                 multihit_model = None
             # Note we are replacing all Ns with As, which means that we need to be careful
             # with masking out these positions later. We do this below.
+            # nt_mask = torch.tensor([it != "N" for it in nt_parent], dtype=torch.bool)
             parent_idxs = sequences.nt_idx_tensor_of_str(nt_parent.replace("N", "A"))
             parent_len = len(nt_parent)
-            molevol.check_csps(parent_idxs, nt_csps)
+            # Cannot assume that nt_csps and mask are same length, because when
+            # datasets are split, masks are recomputed.
+            nt_mask = mask.repeat_interleave(3)[:parent_len]
```
Comment (willdumm) [inline netam/dnsm.py:53]: This is where the comment on splitting datasets came up. For now, I just handle the caveat mentioned in the comment.

File: netam/dxsm.py [hunk 4]
```diff
@@ -55,12 +57,6 @@ def __init__(
         assert len(self.nt_parents) == len(self.nt_children)
         pcp_count = len(self.nt_parents)
 
-        for parent, child in zip(self.nt_parents, self.nt_children):
-            if parent == child:
-                raise ValueError(
-                    f"Found an identical parent and child sequence: {parent}"
-                )
-
```
Comment (willdumm) [inline netam/dxsm.py:63]: This is now done by `assert_pcp_valid` below.

**Comment by willdumm** [comment]
Thanks Erick and Kevin for the helpful feedback!

Just to keep it somewhere, my comment about refactoring `build_selection_matrix_from_parent` expanded:
DXSMBurrito `find_optimal_branch_length` calls
    `self.build_selection_matrix_from_parent`.
    This in turn calls `self.model.selection_factors_of_aa_str`, which takes no
    mask, and computes a mask itself using `netam.common.aa_mask_tensor_of`.
    Shouldn't we be re-using the existing mask instead?
   
On the other hand, as long as we apply any additional mask that we want
        to the result of `build_selection_matrix_from_parent`, the current
        masking method should work fine, since it's just masking ambiguous
        aa sites (minimal masking to get a model output). Maybe this is what we
        want.

**Review by matsen — APPROVED** [review]
Nice! :shipit:
